### PR TITLE
Use specific non-root user to run the container

### LINKED ISSUE #2181: Need to be able to run the container as a non-root user
### Describe the problem and steps to reproduce it:

We need to be able to run the container as a non-root user because the GCP platform doesn't allow running a container as root.

However, currently we can't run the docker image as a non-root user, because the startup script tries to create the `/app/build` directory when the `/app` directory is owned by `root`.

### What happened?
As described above.

### What did you expect to happen?
Be able to run the docker image as a non-root user.

### Anything else we should know?
n/a

### PR DESCRIPTION
**Pull request by bqbn**
The purpose of this PR is to make it possible to run the docker image as a non-root user.

- [X] This PR relates to an existing open issue and there are no existing PRs open for the same issue.
- [X] Add a description of the changes introduced in this PR.
- [X] The change has been successfully run locally.

Fixes #2181

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,26 @@
+FROM node:16 AS builder
+
+WORKDIR /app
+
+# Install the project's dependencies, including dev dependencies.
+COPY package.json yarn.lock ./
+RUN yarn install --pure-lockfile --production=false
+
 FROM node:16
```
Comment (diox) [inline Dockerfile:9]: As above we can probably use the `-slim` image.

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,26 @@
+FROM node:16 AS builder
```
Comment (diox) [inline Dockerfile:1]: We can probably use the `-slim` image, let's do that while we're at it

**Review by diox — COMMENTED** [review]
I get a failure when running it with `docker run -it`:
```
yarn run v1.22.19
$ rm -f .env.local && yarn start-local
$ yarn build && yarn serve
$ react-scripts build --profile
Failed to compile.

EACCES: permission denied, mkdir '/app/build'
```

File: Dockerfile (hunk 2 above)
Comment (bqbn) [reply Dockerfile]: Done.

File: Dockerfile (hunk 1 above)
Comment (bqbn) [reply Dockerfile]: Done.

**Comment by bqbn** [comment]
> I get a failure when running it with `docker run -it`:
> 
> ```
> yarn run v1.22.19
> $ rm -f .env.local && yarn start-local
> $ yarn build && yarn serve
> $ react-scripts build --profile
> Failed to compile.
> 
> EACCES: permission denied, mkdir '/app/build'
> ```

This is a lot like you tried to run as a non-root user using the image built in the `master` branch. If an image is built using the branch of this PR, the `/app` directory would be owned by UID 9500, thus either you run as 9500 or root, you should be able to create the `/app/build` directory.

Locally, I was able to start a container using the image built in this branch, and here is the output:

```
$ docker run --platform linux/amd64 --rm -it -e NODE_ENV=production -e PUBLIC_URL=https://addons-code-manager-dev-cdn.allizom.org -e REACT_APP_API_HOST=https://addons-dev.allizom.org -e REACT_APP_REVIEWERS_HOST=https://reviewers.addons-dev.allizom.org -e REACT_APP_TRACKING_ID=UA-77033033-31 -e SERVER_HOST=0.0.0.0 -e SERVER_PORT=4000 -p 4000 addons-code-manager:latest
yarn run v1.22.19
$ rm -f .env.local && yarn start-local
$ yarn build && yarn serve
$ react-scripts build --profile
Creating an optimized production build...
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Compiled successfully.

File sizes after gzip:

  635.77 KB  build/static/js/2.d25b232b.chunk.js
  37.89 KB   build/static/js/main.c0e5761c.chunk.js
  27.63 KB   build/static/css/main.ba3c5f0f.chunk.css
  1.99 KB    build/static/css/2.6ab7070e.chunk.css
  873 B      build/static/js/runtime-main.d35b9d6c.js

The project was built assuming it is hosted at https://addons-code-manager-dev-cdn.allizom.org/.
You can control this with the homepage field in your package.json.

The build folder is ready to be deployed.

Find out more about deployment here:

  https://cra.link/deployment

$ NODE_ENV=production ts-node -O '{"module": "commonjs", "typeRoots": ["./src/@types"]}' scripts/server.tsx
The node server is listening on http://0.0.0.0:4000 in production mode
```

**Comment by diox** [comment]
```
% git log --oneline -n 1
35f10c2 Use node:16-slim as the base image

% docker build . -t mozilla/addons-frontend:latest
(...)
Successfully built 17c73f5a9279
Successfully tagged mozilla/addons-frontend:latest

% docker images | head -n 2
REPOSITORY                                      TAG                  IMAGE ID       CREATED          SIZE
mozilla/addons-frontend                         latest               17c73f5a9279   11 seconds ago   897MB

% docker run --rm -it mozilla/addons-frontend:latest head -n 1 Dockerfile
FROM node:16-slim AS builder

% docker run --rm -it mozilla/addons-frontend:latest id  
uid=9500(olympia) gid=9500(olympia) groups=9500(olympia)

% docker run --rm -it mozilla/addons-frontend:latest
yarn run v1.22.19
$ rm -f .env.local && yarn start-local
$ yarn build && yarn serve
$ react-scripts build --profile
Failed to compile.

EACCES: permission denied, mkdir '/app/build'


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Inspecting the container, everything inside it has the right owner/group **except** for `/app` which belongs to `root`.

**Comment by bqbn** [comment]
This is what I see locally for a newly built image:

```
$ docker run --platform linux/amd64 --rm -it addons-code-manager:latest ls -ld /app
drwxr-xr-x 1 olympia olympia 4096 Nov 25 06:22 /app
```

What is your docker version? Mine is

```
$ docker --version
Docker version 20.10.16, build aa7e414
```

If the version is not the issue, then I wonder if this is a different behavior on Mac versus Linux.

**Comment by diox** [comment]
I've got `drwxr-xr-x 1 root root 4096 Nov 24 11:39 /app`. Running `Docker version 20.10.12, build 20.10.12-0ubuntu4`.

Related, on a rebuild:
```
% docker build . -t mozilla/addons-frontend:latest
(...)
Step 9/17 : WORKDIR /app
 ---> Using cache
 ---> 59d0fd48b774
Step 10/17 : RUN ln -s /srv/code/version.json /app/version.json
 ---> Running in 57d70e3fb99c
ln: failed to create symbolic link '/app/version.json': Permission denied
The command '/bin/sh -c ln -s /srv/code/version.json /app/version.json' returned a non-zero code: 1
```

Weirdly, I can't build even with `no-cache` now, maybe because of the switch to `slim` ?

```
% docker build . -t mozilla/addons-frontend:latest --no-cache
Sending build context to Docker daemon  454.1MB
Step 1/17 : FROM node:16-slim AS builder
 ---> d0bf2398630e
Step 2/17 : WORKDIR /srv/node
 ---> Running in 5dad54a4642d
Removing intermediate container 5dad54a4642d
 ---> fcc23357d56a
Step 3/17 : COPY package.json yarn.lock /srv/node/
 ---> f753045b2413
Step 4/17 : RUN yarn install --pure-lockfile
 ---> Running in 3111ab5a951d
yarn install v1.22.19
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
[4/5] Linking dependencies...
warning " > react-nested-status@0.2.1" has incorrect peer dependency "react@^16.0.0".
warning "react-nested-status > react-side-effect@1.2.0" has incorrect peer dependency "react@^0.13.0 || ^0.14.0 || ^15.0.0 || ^16.0.0".
warning " > @testing-library/user-event@14.3.0" has unmet peer dependency "@testing-library/dom@>=7.21.4".
warning " > eslint-config-amo@5.1.0" has incorrect peer dependency "jest@^27.0.0".
warning "eslint-config-amo > eslint-config-airbnb@18.2.1" has incorrect peer dependency "eslint@^5.16.0 || ^6.8.0 || ^7.2.0".
warning "eslint-config-amo > eslint-config-airbnb-base@14.2.1" has incorrect peer dependency "eslint@^5.16.0 || ^6.8.0 || ^7.2.0".
warning "eslint-config-amo > eslint-plugin-jsx-a11y@6.4.1" has incorrect peer dependency "eslint@^3 || ^4 || ^5 || ^6 || ^7".
warning "eslint-config-amo > eslint-plugin-react@7.26.1" has incorrect peer dependency "eslint@^3 || ^4 || ^5 || ^6 || ^7".
warning " > eslint-plugin-react-hooks@4.2.0" has incorrect peer dependency "eslint@^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0".
warning "eslint-plugin-testing-library > @typescript-eslint/utils > @typescript-eslint/typescript-estree > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
warning " > po2json@1.0.0-beta-3" has unmet peer dependency "commander@^6.0.0".
warning " > po2json@1.0.0-beta-3" has unmet peer dependency "gettext-parser@2.0.0".
warning " > po2json@1.0.0-beta-3" has unmet peer dependency "gettext-to-messageformat@0.3.1".
warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react@^15.0.0 || ^16.0.0 || ^17.0.0 ".
warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0 || ^17.0.0 ".
warning " > stylelint-config-standard-scss@6.0.0" has incorrect peer dependency "stylelint@^14.14.0".
warning "stylelint-config-standard-scss > stylelint-config-standard@29.0.0" has incorrect peer dependency "stylelint@^14.14.0".
[5/5] Building fresh packages...
info This module is OPTIONAL, you can safely ignore this error
warning Error running install script for optional dependency: "/srv/node/node_modules/unix-dgram: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments: 
Directory: /srv/node/node_modules/unix-dgram
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@9.1.0
gyp info using node@16.18.1 | linux | x64
gyp ERR! find Python 
gyp ERR! find Python Python is not set from command line or npm configuration
gyp ERR! find Python Python is not set from environment variable PYTHON
gyp ERR! find Python checking if \"python3\" can be used
gyp ERR! find Python - \"python3\" is not in PATH or produced an error
gyp ERR! find Python checking if \"python\" can be used
gyp ERR! find Python - \"python\" is not in PATH or produced an error
gyp ERR! find Python 
gyp ERR! find Python **********************************************************
gyp ERR! find Python You need to install the latest version of Python.
gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
gyp ERR! find Python you can try one of the following options:
gyp ERR! find Python - Use the switch --python=\"/path/to/pythonexecutable\"
gyp ERR! find Python   (accepted by both node-gyp and npm)
gyp ERR! find Python - Set the environment variable PYTHON
gyp ERR! find Python - Set the npm configuration variable python:
gyp ERR! find Python   npm config set python \"/path/to/pythonexecutable\"
gyp ERR! find Python For more information consult the documentation at:
gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
gyp ERR! find Python **********************************************************
gyp ERR! find Python 
gyp ERR! configure error 
gyp ERR! stack Error: Could not find any Python installation to use
gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:330:47)
gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:159:21)
gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:202:16)
gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:294:16)
gyp ERR! stack     at exithandler (node:child_process:410:5)
gyp ERR! stack     at ChildProcess.errorhandler (node:child_process:422:5)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
gyp ERR! stack     at onErrorNT (node:internal/child_process:485:16)
gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
gyp ERR! System Linux 5.15.0-52-generic
gyp ERR! command \"/usr/local/bin/node\" \"/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js\" \"rebuild\"
gyp ERR! cwd /srv/node/node_modules/unix-dgram
gyp ERR! node -v v16.18.1
gyp ERR! node-gyp -v v9.1.0
gyp ERR! not ok"
$ husky install
Done in 25.50s.
Removing intermediate container 3111ab5a951d
 ---> db376f8d036a
Step 5/17 : FROM node:16-slim
 ---> d0bf2398630e
Step 6/17 : ARG olympia_uid=9500
 ---> Running in 14106011429f
Removing intermediate container 14106011429f
 ---> 6e6980ff46c7
Step 7/17 : RUN useradd -u ${olympia_uid} -d /home/olympia -m -s /sbin/nologin olympia
 ---> Running in aff5e985f65f
Removing intermediate container aff5e985f65f
 ---> 3af877f3bc0d
Step 8/17 : USER ${olympia_uid}:${olympia_uid}
 ---> Running in d37e7a925ee2
Removing intermediate container d37e7a925ee2
 ---> 6e58aa971061
Step 9/17 : WORKDIR /app
 ---> Running in c51a8f5a869e
Removing intermediate container c51a8f5a869e
 ---> 524867f620a2
Step 10/17 : RUN ln -s /srv/code/version.json /app/version.json
 ---> Running in a4a3bbb29c06
ln: failed to create symbolic link '/app/version.json': Permission denied
The command '/bin/sh -c ln -s /srv/code/version.json /app/version.json' returned a non-zero code: 1
```

**Comment by diox** [comment]
With the following patch:
```diff
diff --git a/Dockerfile b/Dockerfile
index 7ed233bf2..f39020f0b 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,7 @@
 #
 # Build
 #
-FROM node:16-slim AS builder
+FROM node:16 AS builder
 
 WORKDIR /srv/node
 COPY package.json yarn.lock /srv/node/
```
The intermediary error is gone but it still fails with the `Permission Denied` at the end.

**Comment by bqbn** [comment]
Is it possible that you are using the `Dockerfile` in the `addons-frontend` repo to build the image in the `addons-code-manager` repo?

Because the build log in your comment above looks much like the Dockerfile in the `addons-frontend` repo, and this PR is for the `addons-code-manager` repo.

Nonetheless, if you'd like to clear your local env and build from scratch, try run this command and then build in the correct repo.

```
sudo docker system prune -a
```

**Comment by diox** [comment]
Ooops yeah good catch, I had checked out and tried building your `run-as-non-root` branch in addons-frontend as well, and since both the repo and the branch names are so similar that I didn't notice :) Even the tag is wrong!

**Comment by diox** [comment]
So, restarting from scratch: building works, but then running results in the same permission denied error.

```
% docker run --rm -it mozilla/addons-code-manager:latest     
yarn run v1.22.19
$ rm -f .env.local && yarn start-local
$ yarn build && yarn serve
$ react-scripts build --profile
Failed to compile.

EACCES: permission denied, mkdir '/app/build'


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
```
% docker run --rm -it mozilla/addons-code-manager:latest ls -ld /app
drwxr-xr-x 1 root root 4096 Nov 25 20:39 /app
```

**Comment by bqbn** [comment]
I was trying to figure out why using the same Dockerfile, you and I see a different behavior (e.g. my `/app` is owned by 9500, and yours is owned by `root`), and I found this: https://github.com/moby/moby/issues/36677#issuecomment-957357940

It turns out I have `buildkit` enabled by default, how about you?

```
$ ag buildkit .docker/daemon.json
10:    "buildkit": true
```

With that said, the least error-prone way at this point is probably manually set the ownership for the `/app` directory, it seems like.

**Comment by diox** [comment]
It was disabled. Rebuilding with `DOCKER_BUILDKIT=1  docker build . -t mozilla/addons-code-manager:latest` does indeed produce an image with `/app` owned by uid 9500! As you said though we should manually set it to be safe (and try building without `buildkit` just in case)

File: Dockerfile [hunk 3]
```diff
@@ -1,15 +1,42 @@
...
+RUN apt-get update -y && apt-get install -y python3 build-essential
 RUN yarn install --pure-lockfile --production=false
 
-COPY . .
+#
+# Install
+#
+FROM node:16-slim
+
+ARG olympia_uid=9500
```
Comment (diox) [inline Dockerfile:21]: Could we pick another name than `olympia` ? `olympia` is the original name for `addons-server` specifically, so it's a bit weird to use it for a different repos.

File: Dockerfile [hunk 4]
```diff
@@ -1,15 +1,42 @@
...
+ARG app_dir=/app
+
+ENV NODE_ENV production
+ENV SERVER_HOST 0.0.0.0
+ENV PORT 4000
+
+RUN useradd -u ${olympia_uid} -d /home/olympia -m -s /sbin/nologin olympia
```
Comment (diox) [inline Dockerfile:28]: Same as above, I'd prefer if we could avoid `olympia` as the user name.

File: Dockerfile (hunk 3 above)
Comment (bqbn) [reply Dockerfile]: Made it `app_uid`. Hope that is generic enough.

File: Dockerfile (hunk 4 above)
Comment (bqbn) [reply Dockerfile]: Made it `app`.

**Review by diox — APPROVED** [review]

**Pull request merged**